packages/async-flow/src/async-flow.js

```diff
@@ -0,0 +1,491 @@
+            const { bijection, log } = state;
+            const { flow } = facets;
+
+            if (failures.has(flow)) {
+              const prevErr = failures.get(flow);
+              annotateError(
+                prevErr,
+                X`doubly failed somehow with ${fatalProblem}`,
+              );
+              // prevErr likely to be the more relevant diagnostic to report
+              fatalProblem = prevErr;
+            } else {
+              failures.init(flow, fatalProblem);
+            }
+
+            if (hasMembrane(flow)) {
+              getMembrane(flow).stop();
+              deleteMembrane(flow);
+            }
+            log.reset();
+            bijection.reset();
+
+            flow.getFlowState() === 'Failed' ||
+              Fail`Panicked flow must be Failed ${flow}`;
+
+            // This is not an expected throw, so in theory arbitrary chaos
+            // may ensue from throwing it. But at this point
+            // we should have successfully isolated this activation from
+            // having any observable effects on the host, aside from
+            // console logging and
+            // resource exhaustion, including infinite loops
+            const err = makeError(
+              X`In a Failed state: see getFailures() for more information`,
+            );
+            annotateError(err, X`due to ${fatalProblem}`);
+            throw err;
+          },
+        },
+        wakeWatcher: {
+          onFulfilled(_fulfillment) {
+            const { facets } = this;
+            facets.flow.wake();
+          },
+          onRejected(_fulfillment) {
+            const { facets } = this;
+            facets.flow.wake();
+          },
+        },
+      },
+    );
+    const makeAsyncFlowKit = activationArgs => {
+      const asyncFlowKit = internalMakeAsyncFlowKit(activationArgs);
+      const { flow } = asyncFlowKit;
+
+      const vow = toPassableCap(flow.getOutcome());
+      flowForOutcomeVowKey.init(toPassableCap(vow), flow);
+      flow.restart();
+      return asyncFlowKit;
+    };
+    return harden(makeAsyncFlowKit);
+  };
+
+  /**
+   * @param {Zone} zone
+   * @param {string} tag
+   * @param {GuestAsyncFunc} guestFunc
+   * @param {{ startEager?: boolean }} [options]
+   * @returns {HostAsyncFuncWrapper}
+   */
+  const asyncFlow = (zone, tag, guestFunc, options = undefined) => {
+    const makeAsyncFlowKit = prepareAsyncFlowKit(zone, tag, guestFunc, options);
+    const hostFuncName = `${tag}_hostFlow`;
+    const wrapperFunc = {
+      [hostFuncName](...args) {
+        const { flow } = makeAsyncFlowKit(args);
+        return flow.getOutcome();
+      },
+    }[hostFuncName];
+    defineProperties(wrapperFunc, {
+      length: { value: guestFunc.length },
+    });
+    return harden(wrapperFunc);
+  };
+
+  const adminAsyncFlow = outerZone.exo('AdminAsyncFlow', AdminAsyncFlowI, {
+    getFailures() {
+      return failures.snapshot();
+    },
+    wakeAll() {
```

Comment:
I think these methods should be scoped to a flow definition. In particular it's a mistake to call `wakeAll()` before all `asyncFlow` have been redefined.

IMO we should have the following:
- `prepareAsyncFlowTools` would define an exo class kit for flow definitions. Each instance would hold the `failures` and `eagerWakers` stores in their state, and have the following 2 facets:
  - An "admin facet" would implement the `wakeAll()` and `getFailures()` methods. Maybe it could also implement a `getFlowForOutcomeVow()` method (as a shortcut, we may not need to scope to the definition)
  - An "internal facet" allowing the flow kit behavior methods to interact with the stores of the definition as needed.
- `prepareAsyncFlowTools` would also define a heap `WeakMap` mapping async flow (kit) makers to their respective admin facet of the definition instance.
- `prepareAsyncFlowKit` would use a provide pattern for an instance of the above async flow definition kit. On first definition, an instance would be created and stashed somewhere, either in the provided `zone` (likely using a name derived from the flow's tag), or maybe in a shared store in the `outerZone` (if we're ever interested in easily iterating all flow definitions).
  - Technically the kind of the async flow kit could be held in the state of the definition instance, but I don't believe we can make that happen easily with zones.
  - The behavior methods of the flow kit would close over the "internal facet" of the flow definition to interact with the `failures` and `eagerWakers` stores
- `prepareAsyncFlowKit` would register in the WeakMap the "admin facet" of the flow definition to the maker of the flow kit.
  - it will pin the flow definition kit in memory while the maker is retained, but this whole approach already assumes fairly low cardinality of async flow definitions
- `asyncFlow` would similarly associate the admin facet of the flow definition with the `wrapperFunc`
- Finally, the singleton `adminAsyncFlow` can give access to the flow definition's admin facet through the `WeakMap`

This would also allow to automatically `wakeAll` on the flow's definition (similar to how the flow maker "`restart`s" the flow on instantiation.

Reply:
With https://github.com/Agoric/agoric-sdk/issues/9375 filed, to be addressed in a later PR, resolving here.